docs/current/labs/project-zenith.md

```diff
@@ -0,0 +1,539 @@
+
+```sh
+dagger version
+# dagger devel () (jeremyatdockerhub/dagger-engine-worker-zenith) linux/amd64
+```
+
+You can also run a quick hello world to check everything's working:
+
+```sh
+dagger query <<EOF
+{
+  container {
+    from(address:"alpine") {
+      withExec(args:["echo", "hello daggernauts!"]) {
+        stdout
+      }
+    }
+  }
+}
+EOF
+```
+
+### Building from scratch (advanced)
+
+<details>
+<summary>Building from scratch instructions</summary>
+
+To get started, clone or pull this branch:
+
+```sh
+# fresh clone
+git clone https://github.com/shykes/dagger --branch zenith-functions ./dagger/
+
+# OR pull branch
+git remote add shykes https://github.com/shykes/dagger
+git fetch shykes zenith-functions
+git checkout zenith-functions
+```
+
+Next, build the dev `dagger` CLI and start the dev engine:
+
+```sh
+# cd into repo
+cd ./dagger/
+
+# build dev CLI and engine
+./hack/dev
+```
+
+Finally, you need to configure the Dagger environment variables to point to the running engine.
+
+1. If you use [direnv](https://direnv.net/), you can just:
+
+    ```sh
+    cd ./zenith
+    direnv allow .
+    ```
+
+2. If not, you can directly `source` the provided `.envrc` file:
+
+    ```sh
+    cd ./zenith
+    source .envrc
+    ```
+
+At this point you should have a fully functioning `dagger` CLI and dev engine.
+
+You should then be able to use the `dagger` command:
+
+```sh
+./bin/dagger version
+# dagger devel () (registry.dagger.io/engine) linux/amd64
+```
+
+You can also run a quick hello world to check everything's working:
+
+```sh
+./bin/dagger query <<EOF
+{
+  container {
+    from(address:"alpine") {
+      withExec(args:["echo", "hello daggernauts!"]) {
+        stdout
+      }
+    }
+  }
+}
+EOF
+```
+
+</details>
+
+## Creating your first Module
+
+Create a new directory on your filesystem and run `dagger mod init` to
+bootstrap your first module. We'll call it `potato` here, but you can choose
+your favorite food.
+
+```sh
+mkdir potato/
+cd potato/
+
+# initialize Dagger module
+# NOTE: currently Go is the only supported SDK.
+dagger mod init --name=potato --sdk=go
+```
+
+This will generate a `dagger.json` module file, an initial `main.go`
+source file, as well as a generated `dagger.gen.go` and `internal` folder for the generated module code.
+
+If you like, you can run the generated `main.go` like so:
+
+```sh
+echo '{potato{myFunction(stringArg:"Hello daggernauts!"){id}}}' | dagger query
+```
+
+:::note
+All names (functions, arguments, struct fields), etc, are converted into a langauge-agnostic camel-case style.
+:::
+
+Let's try changing the `main.go`. We named our module `potato`, so that means all
+methods on the `Potato` type are published as functions. Let's replace the
+template with something simpler:
+
+```go
+package main
+
+type Potato struct{}
+
+func (m *Potato) HelloWorld() string {
+  return "Hello daggernauts!"
+}
+```
+
+Next, run `dagger mod sync`. **You will need to run this command after every
```

Comment:
Is it still necessary to run this after every change?

Reply:
I think it's only necessary with changes to the "interface" - so if you add new functions, or fields, or change types, etc, then you need it. Otherwise I don't *think* so.